services/hanke-service/src/integrationTest/kotlin/fi/hel/haitaton/hanke/muutosilmoitus/UpdateMuutosilmoitusITest.kt

```diff
@@ -0,0 +1,975 @@
+
+        val exception = assertFailure {
+            muutosilmoitusService.update(muutosilmoitus.id, request, USERNAME)
+        }
+
+        exception.all {
+            hasClass(InvalidHakemusyhteystietoException::class)
+            messageContains("Invalid hakemusyhteystieto received when updating hakemus")
+            messageContains("yhteystietoId=null")
+            messageContains("newId=$requestYhteystietoId")
+        }
+    }
+
+    @ParameterizedTest
+    @EnumSource(ApplicationType::class)
+    fun `throws exception when the request has different persisted contact than the application`(
+        type: ApplicationType
+    ) {
+        val hakemus = hakemusFactory.builder(type).hakija().saveEntity()
+        val muutosilmoitus = muutosilmoitusFactory.builder(hakemus).save()
+        val originalYhteystietoId = muutosilmoitus.hakemusData.yhteystiedot().single().id
+        val requestYhteystietoId = UUID.fromString("ad2173c5-eda4-43e2-8457-7974d74319e8")
+        val request =
+            muutosilmoitus
+                .toUpdateRequest()
+                .withCustomer(CustomerType.COMPANY, requestYhteystietoId)
+
+        val exception = assertFailure {
+            muutosilmoitusService.update(muutosilmoitus.id, request, USERNAME)
+        }
+
+        exception.all {
+            hasClass(InvalidHakemusyhteystietoException::class)
+            messageContains("Invalid hakemusyhteystieto received when updating hakemus")
+            messageContains("yhteystietoId=$originalYhteystietoId")
+            messageContains("newId=$requestYhteystietoId")
+        }
+    }
+
+    @ParameterizedTest
+    @EnumSource(ApplicationType::class)
+    fun `throws exception when the request has a contact that is not a user on hanke`(
+        type: ApplicationType
+    ) {
+        val hakemus = hakemusFactory.builder(type).hakija().saveEntity()
+        val muutosilmoitus = muutosilmoitusFactory.builder(hakemus).save()
+        val originalYhteystietoId = muutosilmoitus.hakemusData.yhteystiedot().single().id
+        val requestHankekayttajaId = UUID.fromString("598e1383-0720-4fd4-8449-21fd759aa457")
+        val request =
+            muutosilmoitus
+                .toUpdateRequest()
+                .withCustomer(CustomerType.COMPANY, originalYhteystietoId, requestHankekayttajaId)
+
+        val exception = assertFailure {
+            muutosilmoitusService.update(muutosilmoitus.id, request, USERNAME)
+        }
+
+        exception.all {
+            hasClass(InvalidHakemusyhteyshenkiloException::class)
+            messageContains("Invalid hanke user/users received when updating hakemus")
+            messageContains("invalidHankeKayttajaIds=[$requestHankekayttajaId]")
+        }
+    }
+
+    @ParameterizedTest
+    @EnumSource(ApplicationType::class)
+    fun `doesn't send email when the caller adds themself as contact`(type: ApplicationType) {
+        val muutosilmoitus = muutosilmoitusFactory.builder().withAreas(listOf()).save()
```

Comment:
Should it be `builder(type)` as in other tests?

Reply:
Yes, thank you.